syntax = "proto3";

package postservice;

message Post {
  string id = 1;
  string title = 2;
  string description = 3;
  string creator_id = 4;
  string created_at = 5;
  string updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
  int64 views = 9;
  int64 likes = 10;
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  string creator_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message CreatePostResponse {
  Post post = 1;
}

message GetPostRequest {
  string post_id = 1;
  string requester_id = 2;
}

message GetPostResponse {
  Post post = 1;
}

message UpdatePostRequest {
  string post_id = 1;
  string requester_id = 2;
  string title = 3;
  string description = 4;
  bool is_private = 5;
  repeated string tags = 6;
}

message UpdatePostResponse {
  Post post = 1;
}

message DeletePostRequest {
  string post_id = 1;
  string requester_id = 2;
}

message DeletePostResponse {
  bool success = 1;
}

message ListPostsRequest {
  string requester_id = 1;
  int32 page = 2;
  int32 page_size = 3;
  string creator_id = 4;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int32 total_count = 2;
}

message ViewPostRequest {
  string post_id = 1;
  string viewer_id = 2;
}

message ViewPostResponse {
  bool success = 1;
}

message LikePostRequest {
  string post_id = 1;
  string user_id = 2;
}

message LikePostResponse {
  bool success = 1;
}

message Comment {
  string id = 1;
  string post_id = 2;
  string user_id = 3;
  string text = 4;
  string created_at = 5;
}

message AddCommentRequest {
  string post_id = 1;
  string user_id = 2;
  string text = 3;
}

message AddCommentResponse {
  Comment comment = 1;
}

message GetCommentsRequest {
  string post_id = 1;
  int32 page = 2;
  int32 page_size = 3;
}

message GetCommentsResponse {
  repeated Comment comments = 1;
  int32 total_count = 2;
}

service PostService {
  rpc CreatePost(CreatePostRequest) returns (CreatePostResponse);
  rpc GetPost(GetPostRequest) returns (GetPostResponse);
  rpc UpdatePost(UpdatePostRequest) returns (UpdatePostResponse);
  rpc DeletePost(DeletePostRequest) returns (DeletePostResponse);
  rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);

  rpc ViewPost(ViewPostRequest) returns (ViewPostResponse);
  rpc LikePost(LikePostRequest) returns (LikePostResponse);
  rpc AddComment(AddCommentRequest) returns (AddCommentResponse);
  rpc GetComments(GetCommentsRequest) returns (GetCommentsResponse);
}
